module/os/linux/zfs/vdev_disk.c

```diff
@@ -916,37 +933,35 @@ vdev_disk_io_rw(zio_t *zio)
 
 	/*
 	 * Check alignment of the incoming ABD. If any part of it would require
-	 * submitting a page that is not aligned to the logical block size,
-	 * then we take a copy into a linear buffer and submit that instead.
-	 * This should be impossible on a 512b LBS, and fairly rare on 4K,
-	 * usually requiring abnormally-small data blocks (eg gang blocks)
-	 * mixed into the same ABD as larger ones (eg aggregated).
+	 * submitting a page that is not aligned to both the logical block size
+	 * and the page size, then we take a copy into a new memory region with
+	 * correct alignment.  This should be impossible on a 512b LBS. On
+	 * larger blocks, this can happen at least when a small number of
+	 * blocks (usually 1) are allocated from a shared slab, or when
+	 * abnormally-small data regions (eg gang headers) are mixed into the
+	 * same ABD as larger allocations (eg aggregations).
 	 */
 	abd_t *abd = zio->io_abd;
-	if (!vdev_disk_check_pages(abd, zio->io_size, bdev)) {
-		void *buf;
-		if (zio->io_type == ZIO_TYPE_READ)
-			buf = abd_borrow_buf(zio->io_abd, zio->io_size);
-		else
-			buf = abd_borrow_buf_copy(zio->io_abd, zio->io_size);
-
+	if (!vdev_disk_check_alignment(abd, zio->io_size, bdev)) {
 		/*
-		 * Wrap the copy in an abd_t, so we can use the same iterators
-		 * to count and fill the vbio later.
+		 * Allocate a new memory region that can guarantee page
+		 * alignment.
 		 */
-		abd = abd_get_from_buf(buf, zio->io_size);
+		abd = abd_alloc_scatter(zio->io_size,
+		    zio->io_abd->abd_flags & ABD_FLAG_META);
```

Comment:
I don't think we really need need abd_alloc_scatter() here or anywhere at all, unless you strictly intend to allocate non-KVA-mapped pages, which abd_alloc() allocates by default for big buffers and which make no sense for small ones.  We should never receive here buffers with size not multiple of ashift, and for those abd_alloc() should always do the right thing.  Or we actually have abd_alloc_for_io(), which should do the best thing for each platform exactly for this use case.

Reply:
I need to guarantee an allocation aligned to the start of a page (KVA-mapped is irrelevant). We don't have a convenient way to get that guarantee in any of our standard allocators.

The problem is that `abd_alloc()` will not necessarily always allocate scatter; it depends on the tuneables that drive `abd_size_alloc_linear()`. For linear, we know that `zio_buf_alloc()` and `zio_data_buf_alloc()` won't necessarily return page-aligned allocations.

If I'm wrong, and we do have an allocator (or set of flags) that is _guaranteed_ to return page-aligned memory, point me at it. I couldn't see it. Otherwise, I have to make one (or adjust an existing one). Options seemed to be:
- round up the allocation size to something we know would get a page-aligned allocation from an existing allocator
- add an explicit "aligned" allocator (or flag to an existing allocator), something like `posix_memalign`
- use `__vmalloc` (that is, `spl_vmalloc()) which appears to guarantee page alignment, but is also fully mapped
- use `alloc_pages()`, which isn't mapped, and is a bit more work

I ended up deciding that `alloc_pages()` is nicer than `__vmalloc()` because apart from the brief moment while copying data from/to the ZIO ABD, those pages don't need to be mapped, and never need to be contiguous. So it should be much kinder for memory pressure. But, it still needs to be wrapped in an ABD, which is more fiddly.

And then I saw `abd_alloc_chunks()` was doing _exactly_ what I needed to do anyway, so I took the shorter route and just exposed its internals.

Though, looking again, I see that we now have `abd_alloc_from_pages()` to wrap an array of page pointers in an ABD, so I guess I could just do it that way (it's very new, so not yet internalised into my standard toolkit heh). I'll give that a go shortly.

I do think that probably `abd_alloc_for_io()` should be trying harder to sort out the proper alignment needed, or pick something that works for everything. Maybe that's not so easy to do if you don't know which device the IO is going to be sent to. I don't know yet. I do think the ABD API is kind of a nightmare; it acts like its abstracting over the details of how a buffer lives in memory, but then in lots of places we have to ask "are you linear?" or "are you scatter?" or whatever. I do intend to try and fix a lot of that, but I don't want to do it here. I freely admit that `abd_alloc_scatter()` is making that mess worse, not better.

Ultimately here, I just need aligned memory. I don't much care where it comes from. I'll try `alloc_pages()` + `abd_alloc_from_pages()` soon. Any other options welcome.

Reply:
You don't need page alignment.  You need a (potentially) virtually contiguous buffer, aligned to ashift (or PAGESIZE if ashift is bigger than it).  And since my last refactor at https://github.com/openzfs/zfs/commit/bd7a02c251d8c119937e847d5161b512913667e6 abd_alloc() (and on Linux abd_alloc_for_io() also as its wrapper) do exactly that, if you provide them with size multiple of ashift, which you will.  The problem workarounded by https://github.com/openzfs/zfs/pull/16678 is that buffer initially allocated as 6KB then truncated to 4KB, that breaks this alignment convention.  If you allocate buffer directly as needed with ashift-multiple size, it will be properly aligned.  Trust me.

Reply:
I suppose no, I don't technically need page-aligned, but that was the only cheap and easy rule to deal with what I was seeing.

So, to be clear: are you saying if I ask for io_size from abd_alloc_for_io, I will never get a LBS block crossing a memory page? If so, very excellent, and I'm gonna add commentary and assertions to that effect so I never have to experience this ever again 😆

Reply:
Yes, that is what I am saying. Obviously while LBS <= PAGESIZE.  I remember I already wondered once what will Linux do when disks with LBS > PAGESIZE appear one day.

Reply:
I'm sure "cry".

Last push removed `abd_alloc_scatter()` and switched to `abd_alloc_for_io()`. Seems ok!